<a href="https://colab.research.google.com/github/vidjaydas/Data_Science_Python/blob/main/EDA_2_Iris_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### As a part of EDA-2 we have discussed

1.   Label Encoder
2.   One Hot Encoder - using Sklearn & Pandas

3.   Isolation Forest
4.   PPS

#Label Encoder

In [1]:
from sklearn.preprocessing import LabelEncoder
import pandas as pd
import seaborn as sns

In [2]:
data = sns.load_dataset('iris')

In [3]:
data.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [4]:
#Change does not effect the "data" dataframe
data1 = data.copy()

In [5]:
labelencoder = LabelEncoder()
data1.iloc[:,-1] = labelencoder.fit_transform(data.iloc[:,-1])

In [6]:
data1

sepal_length  sepal_width  petal_length  petal_width  species
0             5.1          3.5           1.4          0.2        0
1             4.9          3.0           1.4          0.2        0
2             4.7          3.2           1.3          0.2        0
3             4.6          3.1           1.5          0.2        0
4             5.0          3.6           1.4          0.2        0
..            ...          ...           ...          ...      ...
145           6.7          3.0           5.2          2.3        2
146           6.3          2.5           5.0          1.9        2
147           6.5          3.0           5.2          2.0        2
148           6.2          3.4           5.4          2.3        2
149           5.9          3.0           5.1          1.8        2

[150 rows x 5 columns]

LabelEncoder is used on categorical variable to encode it to classification

#One Hot Encoder (OHE)

#Using Sklearn
it is a very lenthy process. it is very easy to perform using pandas

In [7]:
from sklearn.preprocessing import OneHotEncoder

In [8]:
data2 = sns.load_dataset('iris')

In [9]:
data2

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

In [10]:
# Creating instance of one-hot-encoder
enc = OneHotEncoder(handle_unknown='ignore')

In [11]:
#Passing bridge-types-cat column (label encoded)
enc_df = pd.DataFrame(enc.fit_transform(data2[['species']]).toarray())     #toarray() means we create an array and store the data in it

In [12]:
enc_df

0    1    2
0    1.0  0.0  0.0
1    1.0  0.0  0.0
2    1.0  0.0  0.0
3    1.0  0.0  0.0
4    1.0  0.0  0.0
..   ...  ...  ...
145  0.0  0.0  1.0
146  0.0  0.0  1.0
147  0.0  0.0  1.0
148  0.0  0.0  1.0
149  0.0  0.0  1.0

[150 rows x 3 columns]

In [13]:
# Merge with main df
data_final = data2.iloc[:, 0:4].join(enc_df)
data_final

sepal_length  sepal_width  petal_length  petal_width    0    1    2
0             5.1          3.5           1.4          0.2  1.0  0.0  0.0
1             4.9          3.0           1.4          0.2  1.0  0.0  0.0
2             4.7          3.2           1.3          0.2  1.0  0.0  0.0
3             4.6          3.1           1.5          0.2  1.0  0.0  0.0
4             5.0          3.6           1.4          0.2  1.0  0.0  0.0
..            ...          ...           ...          ...  ...  ...  ...
145           6.7          3.0           5.2          2.3  0.0  0.0  1.0
146           6.3          2.5           5.0          1.9  0.0  0.0  1.0
147           6.5          3.0           5.2          2.0  0.0  0.0  1.0
148           6.2          3.4           5.4          2.3  0.0  0.0  1.0
149           5.9          3.0           5.1          1.8  0.0  0.0  1.0

[150 rows x 7 columns]

#Using Pandas

In [14]:
import pandas as pd

In [15]:
data3 = sns.load_dataset('iris')

In [17]:
data_encoded = pd.get_dummies(data3)
data_encoded

sepal_length  sepal_width  ...  species_versicolor  species_virginica
0             5.1          3.5  ...                   0                  0
1             4.9          3.0  ...                   0                  0
2             4.7          3.2  ...                   0                  0
3             4.6          3.1  ...                   0                  0
4             5.0          3.6  ...                   0                  0
..            ...          ...  ...                 ...                ...
145           6.7          3.0  ...                   0                  1
146           6.3          2.5  ...                   0                  1
147           6.5          3.0  ...                   0                  1
148           6.2          3.4  ...                   0                  1
149           5.9          3.0  ...                   0                  1

[150 rows x 7 columns]

#IsolationForest

In [18]:
from sklearn.ensemble import IsolationForest

In [19]:
data = sns.load_dataset('iris')
data_encoded = pd.get_dummies(data)

In [21]:
#Training the model
clf = IsolationForest(random_state=10, contamination=0.01)  #Contamination meaning based on the domain knowledge how many numbers if outliers can be possible in the dataset
clf.fit(data_encoded)

IsolationForest(behaviour='deprecated', bootstrap=False, contamination=0.01,
                max_features=1.0, max_samples='auto', n_estimators=100,
                n_jobs=None, random_state=10, verbose=0, warm_start=False)

In [22]:
# Predictions
y_pred_outliers = clf.predict(data_encoded)

In [23]:
# -1 for outliers and 1 for inliers
y_pred_outliers

array([ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1, -1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1, -1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1])

In [25]:
data_encoded.head()

sepal_length  sepal_width  ...  species_versicolor  species_virginica
0           5.1          3.5  ...                   0                  0
1           4.9          3.0  ...                   0                  0
2           4.7          3.2  ...                   0                  0
3           4.6          3.1  ...                   0                  0
4           5.0          3.6  ...                   0                  0

[5 rows x 7 columns]

In [27]:
# Let us add a new data point which is outlier
data_encoded.loc[150] = [20, 40, 30, 50, 1, 0, 0]

In [28]:
data_encoded

sepal_length  sepal_width  ...  species_versicolor  species_virginica
0             5.1          3.5  ...                   0                  0
1             4.9          3.0  ...                   0                  0
2             4.7          3.2  ...                   0                  0
3             4.6          3.1  ...                   0                  0
4             5.0          3.6  ...                   0                  0
..            ...          ...  ...                 ...                ...
146           6.3          2.5  ...                   0                  1
147           6.5          3.0  ...                   0                  1
148           6.2          3.4  ...                   0                  1
149           5.9          3.0  ...                   0                  1
150          20.0         40.0  ...                   0                  0

[151 rows x 7 columns]

In [30]:
# Training the model
clf = IsolationForest(random_state=10, contamination=0.01)
clf.fit(data_encoded)

# Predictions
y_pred_outliers = clf.predict(data_encoded)
y_pred_outliers

array([ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1, -1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1, -1])

In [31]:
data_encoded['scores'] = clf.decision_function(data_encoded)

In [32]:
data_encoded['anamoly'] = clf.predict(data_encoded.iloc[:, 0:7])

In [33]:
data_encoded

sepal_length  sepal_width  ...    scores  anamoly
0             5.1          3.5  ...  0.201438        1
1             4.9          3.0  ...  0.183516        1
2             4.7          3.2  ...  0.173272        1
3             4.6          3.1  ...  0.181867        1
4             5.0          3.6  ...  0.198218        1
..            ...          ...  ...       ...      ...
146           6.3          2.5  ...  0.134374        1
147           6.5          3.0  ...  0.178850        1
148           6.2          3.4  ...  0.122817        1
149           5.9          3.0  ...  0.156061        1
150          20.0         40.0  ... -0.287729       -1

[151 rows x 9 columns]

In [34]:
#Print the putliers data points
data_encoded[data_encoded['anamoly']==-1]

sepal_length  sepal_width  ...    scores  anamoly
106           4.9          2.5  ... -0.008763       -1
150          20.0         40.0  ... -0.287729       -1

[2 rows x 9 columns]

This are the outloers, if we examine the 106 data point the values are similar to the other datapoints, so we can decrease the contamination value.

#PPS Score

In [35]:
# Install the package
!pip install ppscore

     |████████████████████████████████| 51kB 1.6MB/s 
  Created wheel for ppscore: filename=ppscore-1.2.0-py2.py3-none-any.whl size=13050 sha256=283d614189cd5331fab5fcdab36a2a367397093e52db34827f105f45d24d7f47
  Stored in directory: /root/.cache/pip/wheels/50/cc/8f/d0da07e4ec4e7a4c4cf32aee9057e4d1c318de910543b4684e
Successfully built ppscore


In [36]:
import ppscore as pps

In [40]:
data3.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')

In [41]:
#pps.score(df, 'feature_column', 'target_column') syntax
pps.score(data3, 'sepal_length', 'petal_length')

{'baseline_score': 1.4886666666666668,
 'case': 'regression',
 'is_valid_score': True,
 'metric': 'mean absolute error',
 'model': DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best'),
 'model_score': 0.6692708968366863,
 'ppscore': 0.550422595049248,
 'x': 'sepal_length',
 'y': 'petal_length'}

In [42]:
#Calculate the whole PPS matrix
pps.matrix(data3)

x  ...                                              model
0   sepal_length  ...                                               None
1   sepal_length  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
2   sepal_length  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
3   sepal_length  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
4   sepal_length  ...  DecisionTreeClassifier(ccp_alpha=0.0, class_we...
5    sepal_width  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
6    sepal_width  ...                                               None
7    sepal_width  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
8    sepal_width  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
9    sepal_width  ...  DecisionTreeClassifier(ccp_alpha=0.0, class_we...
10  petal_length  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
11  petal_length  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
12  petal_length  ...                                               None
13  petal_length  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
14  petal_length  ...  DecisionTreeClassifier(ccp_alpha=0.0, class_we...
15   petal_width  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
16   petal_width  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
17   petal_width  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
18   petal_width  ...                                               None
19   petal_width  ...  DecisionTreeClassifier(ccp_alpha=0.0, class_we...
20       species  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
21       species  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
22       species  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
23       species  ...  DecisionTreeRegressor(ccp_alpha=0.0, criterion...
24       species  ...                                               None

[25 rows x 9 columns]

if the pps score is high / close to 1, then we can say that the pair are strongly associated.